### PR TITLE
Modernize Systolic GEMM and fix issues along the way

### PR DESCRIPTION
**Pull request by definelicht**
Modernize the systolic GEMM sample, and fix a handful of issues that popped up along the way:
- Handling of `StreamArray` in Xilinx wrongly removed the index
- Stream buffer sizes was not being symbolically evaluated before checking if the value is compile-time fixed
- Processing element unrolling for Intel FPGA made a wrong assumption
- Intel codegen should actually use the memlet name when generating channel accesses
- Intel codegen was defining memlets accessing stream arrays as stream arrays, when they should be streams
- Added missing undefines causing warnings

### CODE REVIEW

File: dace/codegen/targets/intel_fpga.py [hunk 1]
```diff
@@ -538,40 +536,39 @@ def generate_module(self, sdfg, state, name, subgraph, parameters,
...
         # Add kernel call host function
-        if unrolled_loops == 0:
+        if unrolled_loop is None:
             host_body_stream.write(
                 "kernels.emplace_back(program.MakeKernel(\"{}\"{}));".format(
                     module_function_name, ", ".join([""] + kernel_args_call)
                     if len(kernel_args_call) > 0 else ""), sdfg, state_id)
         else:
             # We will generate a separate kernel for each PE. Adds host call
-            for ul in self._unrolled_pes:
```
Comment (TizianoDeMatteis) [inline dace/codegen/targets/intel_fpga.py:555]: Can't we have nested maps that generates PE?

File: dace/codegen/targets/intel_fpga.py [hunk 2]
```diff
@@ -257,8 +257,6 @@ def make_read(self, defined_type, dtype, var_name, expr, index, is_pack,
         if defined_type == DefinedType.Stream:
             read_expr = "read_channel_intel({})".format(expr)
         elif defined_type == DefinedType.StreamArray:
-            # remove "[0]" index as this is not allowed if the subscripted value is not an array
-            expr = expr.replace("[0]", "")
```
Comment (TizianoDeMatteis) [inline dace/codegen/targets/intel_fpga.py:261]: Is this handled differently?
IIRC there was a case where a single stream was accessed as an array.
(But this wasn't the right place to fix it)

Comment (definelicht) [reply dace/codegen/targets/intel_fpga.py:261]: This was a bad hack that covered up the actual problem, which was that we defined memlets `#define pipe pipes[i]` as a `StreamArray`, when it should be a `Stream`. This is fixed now, so this hack is no longer necessary :-)

File: dace/codegen/targets/intel_fpga.py (hunk 1 above)
Comment (definelicht) [reply dace/codegen/targets/intel_fpga.py:555]: Maybe, but the rest of the code you wrote assumes maximum one map :P

**Review by TizianoDeMatteis — APPROVED** [review]

**Pull request merged**